Big Bang Component Extension (#1094)
defenseunicorns/zarf: docs/13-walkthroughs/5-big-bang.md

# Using Big Bang with Zarf

## Introduction

This walkthrough describes how to use Big Bang with Zarf for Air Gap deployments through the use of the Big Bang Zarf extension. If you are not familiar with Big Bang you can learn more about it here: https://p1.dso.mil/products/big-bang, but in short it is a DevSecOps platform that contains many useful tools for building, managing, and running software projects while adhering to the [United States Department of Defense DevSecOps Reference Design](https://public.cyber.mil/devsecops/).

Zarf integrates with Big Bang through the use of an extension that simplifies the selection of Big Bang packages and the pulling of the required artifacts to deploy those packages in an Air Gap.

### Known Issues

> The current version of this extension requires [some patches](../../examples/big-bang/config/patch-images.yaml) to the Big Bang values to function correctly, and is not fully integrated into the `zarf package remove` lifecycle.  We will be looking to remove these limitations in a future release.

Reviewer: ```suggestion
The current version of this extension works best with Big Bang version `1.54.0` or later, and is not fully integrated into the `zarf package remove` lifecycle.  We will be looking to remove these limitations in a future release.
```

## System Requirements

Big Bang requires a reasonably powerful `amd64` system that scales up with the number of components deployed.  We recommend at least `32 GB` of RAM and a high-speed internet connection to complete this walkthrough.

> To learn more about Big Bang's requirements in general, see their documentation: https://docs-bigbang.dso.mil/1.52.0/docs/prerequisites/minimum-hardware-requirements/

Reviewer 2: The docs will re-direct to the latest tag using `latest`, so a potentially better URL here is <https://docs-bigbang.dso.mil/latest/docs/prerequisites/minimum-hardware-requirements/>

:::

## Package Creation

The below section covers creating and tuning the Big Bang package:

### Setup

By default, Big Bang uses images from [Iron Bank](https://p1.dso.mil/products/iron-bank) which will require you to set your login credentials for [Registry One](https://registry1.dso.mil) (see [pre-requisites](#pre-requisites) for information on account setup).

```bash
# Authenticate to https://registry1.dso.mil/, then retrieve your CLI secret from your User Profile and run the following:
set +o history
export REGISTRY1_USERNAME=<REPLACE_ME>
export REGISTRY1_CLI_SECRET=<REPLACE_ME>
echo $REGISTRY1_CLI_SECRET | zarf tools registry login registry1.dso.mil --username $REGISTRY1_USERNAME --password-stdin
set -o history
```

Now navigate to the `examples/big-bang` folder within the Zarf repository you cloned in the [pre-requisites](#pre-requisites) section.

### Configure Big Bang

Within the `examples/big-bang` folder you will see a `zarf.yaml` that has the following [component](../4-user-guide/2-zarf-packages/2-zarf-components.md) defined:

```yaml
components:
  - name: bigbang
    required: true
    extensions:
      bigbang:

>         version: 1.52.0
>         skipFlux: false
>         valuesFiles:
>           - config/minimal.yaml #turns on just istio
>           - config/ingress.yaml # adds istio certs for *.bigbang.dev
>           - config/kyverno.yaml # turns on kyverno
>           - config/loki.yaml # turns on loki and monitoring
>           - config/patch-images.yaml # fixes upstream bug in BB

Reviewer: ```suggestion
        version: 1.54.0
        skipFlux: false
        valuesFiles:
          - config/minimal.yaml #turns on just istio
          - config/ingress.yaml # adds istio certs for *.bigbang.dev
          - config/kyverno.yaml # turns on kyverno
          - config/loki.yaml # turns on loki and monitoring
```

```

This component uses the `bigbang` extension to define the version of Big Bang to use and the values files to apply.  Feel free to inspect and configure the values.yaml files as you wish and to learn more about Big Bang's configuration see their values guide: https://docs-bigbang.dso.mil/latest/docs/guides/using-bigbang/values-guide/

:::note

The `valuesFiles` are applied from top to bottom and will apply the last value that was provided for any given key.

:::

:::note

> Versions of Big Bang prior to `1.54.0` require the `config/patch-images.yaml` values file to be included.

Reviewer: ```suggestion
This extension works best with Big Bang version `1.54.0` or later.  Version `1.53.0` requires manual patches to images to function correctly.
```

```bash
zarf package create
```

Now wait for the package creation to complete and you should see a `zarf-package-big-bang-example-amd64-x.x.x.tar.zst` file in the directory.

## Package Deployment

The below section covers deploying the Big Bang package from the previous section:

### Initialize Zarf

Before you can deploy the Big Bang package you must first initialize Zarf on the cluster you created in the [pre-requisites](#pre-requisites) section.  To do so you can run the following:

```bash
# Initialize Zarf (interactively)
zarf init
# Make these choices at the prompts
# ? Do you want to download this init package? Yes
# ? Deploy this Zarf package? Yes
# ? Deploy the k3s component? No
# ? Deploy the logging component? No
# ? Deploy the git-server component? Yes

# (Optional) Inspect the results
zarf tools k9s
```

:::note

The `git-server` component is required by Big Bang as it uses it as a source for Flux deployments.

:::

### Deploy Big Bang

Now you are ready to deploy Big Bang, and can do so with the following in the `examples/big-bang` directory:

```bash
# Deploy Big Bang (interactively)

> ./zarf package deploy

Reviewer 2: ```suggestion
zarf package deploy
```

# Make these choices at the prompts
# ? Choose or type the package file [tab for suggestions] zarf-package-big-bang-example-amd64-x.x.x.tar.zst
# ? Deploy this Zarf package? Yes
```

> ### See the results

Reviewer 2: ```suggestion
### See The Results
```

All other headers are fully capitalized, this is something that would be good to define in our docs style guide @Madeline-UX

Reviewer 3: Agreed - Will add all headers to be in title caps to docs style guide
